### PR TITLE
Power Monitoring Console

### PR DESCRIPTION
**Pull request by 20kdc**
## About the PR

This is mostly a port of an ancient test thing, so that these consoles show *something*.

Presumably, someone will at some point work out how much exactly these ought to show, but for now they show total incoming and outgoing power, along with a breakdown by device on the HV network.

Assuming that substations get named accordingly at some point, this allows for a detailed breakdown of how much power is being used per-department.

Note that due to the weirdness of battery logic source values may act a bit funny around them. Not sure how to deal with this.

Also note that since this is a port of something from a year ago, I may have made some mistakes during the porting process.

**Screenshots**
image

**Changelog**

:cl:
- add: Power Monitoring Consoles now show some power details on high-voltage networks.

### CODE REVIEW

**Comment by HoofedEar** [comment]
Would it be too based if these showed the coordinates of the substation/SMES ?

File: Content.Server/Power/Components/PowerMonitoringConsoleComponent.cs [hunk 1]
```diff
@@ -0,0 +1,17 @@
+#nullable enable
```
Comment (mirrorcult) [inline Content.Server/Power/Components/PowerMonitoringConsoleComponent.cs:1]:
```suggestion
```

File: Content.Server/Power/EntitySystems/PowerMonitoringConsoleSystem.cs [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+internal sealed class PowerMonitoringConsoleSystem : EntitySystem
+{
+    private float _updateTimer = 0.0f;
+
+    public override void Update(float frameTime)
+    {
+        _updateTimer += frameTime;
+        if (_updateTimer >= 1)
+        {
+            _updateTimer -= 1;
```
Comment (mirrorcult) [inline Content.Server/Power/EntitySystems/PowerMonitoringConsoleSystem.cs:23]: the 1 here should be a static/const in the system

File: Content.Server/Power/EntitySystems/PowerMonitoringConsoleSystem.cs [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+        if (netQ != null)
+        {
+            var net = netQ!;
+            foreach (PowerConsumerComponent pcc in net.Consumers)
+            {
+                if (pcc.DrawRate <= 0)
+                {
+                    // Grilles & other inactive power consumers.
+                    continue;
+                }
```
Comment (mirrorcult) [inline Content.Server/Power/EntitySystems/PowerMonitoringConsoleSystem.cs:59]: i feel like there should be an option/checkbox to show inactive power consumers? could be useful

File: Content.Server/Power/EntitySystems/PowerMonitoringConsoleSystem.cs [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+                var rate = batteryComp.NetworkBattery.CurrentSupply;
+                sources.Add(LoadOrSource(pcc, rate));
+                totalSources += rate;
+            }
+        }
+        // Sort
+        loads.Sort(CompareLoadOrSources);
+        sources.Sort(CompareLoadOrSources);
+        // Actually set state.
+        target.GetUIOrNull(PowerMonitoringConsoleUiKey.Key)?.SetState(new PowerMonitoringConsoleBoundInterfaceState(totalSources, totalLoads, sources.ToArray(), loads.ToArray()));
```
Comment (mirrorcult) [inline Content.Server/Power/EntitySystems/PowerMonitoringConsoleSystem.cs:93]: use the `UserInterfaceSystem` methods, the entityuid extension for `GetUIOrNull` is quite slow and deprecated

File: Resources/Prototypes/Entities/Structures/Machines/Computers/computers.yml [hunk 5]
```diff
@@ -75,6 +75,20 @@
     color: "#c9c042"
   - type: Computer
     board: PowerComputerCircuitboard
+  - type: PowerMonitoringConsole
+    voltage: High
```
Comment (mirrorcult) [inline Resources/Prototypes/Entities/Structures/Machines/Computers/computers.yml:79]: this doesnt seem to be an actual field on this component?

**Review by mirrorcult — CHANGES_REQUESTED** [review]

File: Resources/Prototypes/Entities/Structures/Machines/Computers/computers.yml (hunk 5 above)
Comment (20kdc) [reply Resources/Prototypes/Entities/Structures/Machines/Computers/computers.yml]: Fixed in ac07a2b5d932ee335d2eb99c6105a3eec0024cb6

File: Content.Server/Power/EntitySystems/PowerMonitoringConsoleSystem.cs (hunk 4 above)
Comment (20kdc) [reply Content.Server/Power/EntitySystems/PowerMonitoringConsoleSystem.cs]: Fixed in ac07a2b5d932ee335d2eb99c6105a3eec0024cb6

File: Content.Server/Power/Components/PowerMonitoringConsoleComponent.cs (hunk 1 above)
Comment (20kdc) [reply Content.Server/Power/Components/PowerMonitoringConsoleComponent.cs]: As I said, ported old test code...
Anyway fixed in ac07a2b5d932ee335d2eb99c6105a3eec0024cb6

File: Content.Server/Power/EntitySystems/PowerMonitoringConsoleSystem.cs (hunk 2 above)
Comment (20kdc) [reply Content.Server/Power/EntitySystems/PowerMonitoringConsoleSystem.cs]: Fixed in ac07a2b5d932ee335d2eb99c6105a3eec0024cb6

File: Content.Server/Power/EntitySystems/PowerMonitoringConsoleSystem.cs (hunk 3 above)
Comment (20kdc) [reply Content.Server/Power/EntitySystems/PowerMonitoringConsoleSystem.cs]: Added as of 3823be9b1

**Review by mirrorcult — APPROVED** [review]

**Pull request merged**